lib/chef/knife/bmcs_helper.rb

```diff
@@ -57,43 +57,99 @@ def warn_if_page_is_truncated(response)
         ui.warn('This list has been truncated. To view more items, increase the limit.') if response.headers.include? 'opc-next-page'
       end
 
-      # TODO: Method should be refactored to reduce complexity.
-      # rubocop:disable Metrics/PerceivedComplexity
-      def display_list(response, columns, warn_on_truncated: true)
-        list = if response.data.nil?
-                 []
-               else
-                 response.data.is_a?(Array) ? response.data : [response.data]
-               end
+      def next_page_token(response)
+        return response.headers['opc-next-page'] if response.headers.include? 'opc-next-page'
+        nil
+      end
+
+      def get_display_results(options)
+        max_results = config[:limit] ? Integer(config[:limit]) : nil
+
+        num_fetched_results = 0
         list_for_display = []
+        first_row = true
+        response = nil
+        loop do
+          response, new_items = yield(options, first_row)
+
+          list_for_display += new_items
+          num_fetched_results += response.data.length if response.data
+          break if next_page_token(response).nil?
+          break if max_results && num_fetched_results >= max_results
+          first_row = false
+          options[:page] = next_page_token(response)
+          options[:limit] = (max_results - num_fetched_results).to_s if max_results
+        end
+        [list_for_display, response]
+      end
 
-        if config[:format] == 'summary'
-          width = 1
+      # Return data in summary mode format, optionally including column headings.
+      def _summary_list(list, columns, include_headings: true)
```

Comment:
Just a suggestion, but would it make sense to add the headers first before anything else so that then columns and include_headers don't have to be passed around so much?

Reply:
Passing in the same args repeatedly often indicates something should be made into a class of its own.  I'll give it some thought.
On unit test coverage, coverage was the same except for an unused method which I will remove.
On integration testing, server-side behaviour is not predictable with respect to when 'partial' replies are given.  I'm open to suggestions.